core-java-modules/core-java-io-5/src/main/java/com/baeldung/usinggzipInputstream/Main.java

```diff
@@ -0,0 +1,41 @@
+package com.baeldung.usinggzipInputstream;
+
+import java.io.*;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.function.Consumer;
+import java.util.zip.GZIPInputStream;
+
+public class Main {
+    static String filePath = Objects.requireNonNull(Main.class.getClassLoader().getResource("myFile.gz")).getFile();
+
+    public static void main(String[] args) throws IOException {
+
+        try {
+            readGZipFileWithFilter(filePath, lines -> lines.forEach(System.out::println));
+        } catch (FileNotFoundException e) {
+            System.err.println(e.getMessage());
+        }
+    }
+
+    public static List<String> readGZipFile(String filePath) throws IOException {
+        List<String> lines = new ArrayList<>();
+        try (InputStream inputStream = new FileInputStream(filePath);
+             GZIPInputStream gzipInputStream = new GZIPInputStream(inputStream);
+             InputStreamReader inputStreamReader = new InputStreamReader(gzipInputStream);
+             BufferedReader bufferedReader = new BufferedReader(inputStreamReader)) {
+
+            String line;
+            while ((line = bufferedReader.readLine()) != null) {
+                lines.add(line);
+            }
+        }
+
+        return lines;
+    }
```

Comment:
This algorithm is good. I also expect to see two more variations:

```java
    public static List<String> findInZipFile(String filePath, String toFind) throws IOException {
        try (InputStream inputStream = new FileInputStream(filePath);
             GZIPInputStream gzipInputStream = new GZIPInputStream(inputStream);
             InputStreamReader inputStreamReader = new InputStreamReader(gzipInputStream);
             BufferedReader bufferedReader = new BufferedReader(inputStreamReader)) {

             return bufferedReader.lines().filter(line -> line.contains(toFind)).collect(toList());
        }
    }

    public static List<String> useContentsOfZipFile(String filePath, Consumer<Stream<String>> consumer) throws IOException {
        try (InputStream inputStream = new FileInputStream(filePath);
             GZIPInputStream gzipInputStream = new GZIPInputStream(inputStream);
             InputStreamReader inputStreamReader = new InputStreamReader(gzipInputStream);
             BufferedReader bufferedReader = new BufferedReader(inputStreamReader)) {

            consumer.accept(bufferedReader.lines());
        }
    }


```

We can also unit test these. We explain how it's important to provide the `try-with-resources` in each case, hence needing the `lines` to be used within that construct.